collab-service/src/tunneller/collab_tunnel_controller.ts

```diff
@@ -1,40 +1,65 @@
 import { ServerDuplexStream } from '@grpc/grpc-js';
 import { createClient, RedisClientType } from 'redis';
-import { createDisconnectMessage } from '../room/disconnect_message_builder';
-import { createRedisPubSubAdapter } from '../redis_adapter/redis_pubsub_adapter';
+import { createAckMessage, createJoinMessage } from '../message_handler/internal/internal_message_builder';
+import { createConnectedMessage, createDisconnectedMessage } from '../message_handler/room/connect_message_builder';
+import { createRedisPubSubAdapter, TunnelPubSub } from '../redis_adapter/redis_pubsub_adapter';
 import { createRedisTopicPool, RedisTopicPool } from '../redis_adapter/redis_topic_pool';
 import createRoomSessionService from '../room_auth/room_session_agent';
-import createUnauthorizedMessage from '../room/unauthorized_message_builder';
+import createUnauthorizedMessage from '../message_handler/room/unauthorized_message_builder';
 import createQuestionService from '../question_client/question_agent';
 import setQuestionRedis from '../redis_adapter/redis_question_adapter';
-import { CollabTunnelRequest, CollabTunnelResponse, VerifyRoomErrorCode } from '../proto/collab-service';
-import { CollabTunnelSerializer, TunnelMessage } from './collab_tunnel_serializer';
+import Logger from '../utils/logger';
+import CollabTunnelSerializer from './collab_tunnel_serializer';
+import {
+  CollabTunnelRequest,
+  CollabTunnelResponse,
+  VerifyRoomErrorCode,
+} from '../proto/collab-service';
 import { IRoomSessionAgent } from '../room_auth/room_session_agent_types';
 import { IQuestionAgent } from '../question_client/question_agent_types';
+import { ConnectionFlag, TunnelMessage } from '../message_handler/internal/internal_message_types';
 
 const PROXY_HEADER_USERNAME = 'X-Gateway-Proxy-Username';
 const PROXY_HEADER_NICKNAME = 'X-Gateway-Proxy-Nickname';
 const PROXY_HEADER_ROOM_TOKEN = 'X-Gateway-Proxy-Room-Token';
 
 function createCallWriter(
   call: ServerDuplexStream<CollabTunnelRequest, CollabTunnelResponse>,
+  pubsub: TunnelPubSub<TunnelMessage>,
   username: string,
+  nickname: string,
 ): (data: TunnelMessage) => void {
-  return (message: TunnelMessage): void => {
-    const res = CollabTunnelResponse.create(
-      {
-        data: Buffer.from(message.data),
-        flags: VerifyRoomErrorCode.VERIFY_ROOM_ERROR_NONE,
-      },
-    );
-
+  // Creates collab response to be sent to client
+  function makeResponse(data: Uint8Array): CollabTunnelResponse {
+    return CollabTunnelResponse.create({
+      data: Buffer.from(data),
+      flags: VerifyRoomErrorCode.VERIFY_ROOM_ERROR_NONE,
+    });
+  }
+  return async (message: TunnelMessage): Promise<void> => {
     if (message.sender !== username) {
```

Comment:
Minor nitpick: Guard clause can be improved
```suggestion
    if (message.sender === username) {
        return;
    }
    // rest of code here
```